docs/architecture/adr-056-orm-query.md

```diff
@@ -0,0 +1,123 @@
+# ADR 056: ORM Query Layer
+
+## Changelog
+
+* {date}: {changelog}
+
+## Status
+
+{DRAFT | PROPOSED} Not Implemented
+
+> Please have a look at the [PROCESS](./PROCESS.md#adr-status) page.
+> Use DRAFT if the ADR is in a draft stage (draft PR) or PROPOSED if it's in review.
+
+## Abstract
+
+> "If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the ADR.
+> A short (~200 word) description of the issue being addressed.
+
+## Context
+
+> This section describes the forces at play, including technological, political, social, and project local. These forces are probably in tension, and should be called out as such. The language in this section is value-neutral. It is simply describing facts. It should clearly explain the problem and motivation that the proposal aims to resolve.
+> {context body}
+
+## Decision
+
+> This section describes our response to these forces. It is stated in full sentences, with active voice. "We will ..."
+> {decision body}
+
+### Logical Query Expressions
+
+Internally queries will be processed using logical query expressions. This will allow queries from the different APIs
+to target any backends that support these logical query expressions. Both a simple generic gRPC API and a type-safe
+GraphQL API are proposed as well as ORM state store and SQL backends.
+
+A query expression can specify:
+* the message name of the table being queried
+* a where expression, explained below,
+* an order by expression which includes 0 or more fields in either ascending or descending order
+* a pagination request (`cosmos.base.query.v1beta1.PageQuest`)
+
+#### Where Expressions
+
+A where expression is composed of one or more field expressions. Each field expression may specify one or two
+comparison operators on that field, either:
+* `eq`, or
+* `lt` or `lte` &/or `gt` or `gte`
+
+Other comparison operators may be added in the future such as `neq`, `in`, `nin`, `regex`, etc.
+More generic `and` and `or` expressions may also be added in the future, but for
+now the goal is to keep the query layer as simple as possible so that the most value
+can be provided with the least effot. Between the two goals of improving the query language and
+adding support for more backends and improving performance, the latter should
+be preferred.
+
+### Built-in Query Planner
+
+The built-in query planner which targets ORM state directly should be as simple as possible without being 100%
+naive. The proposed approach is to use a simple rule-based query planner
+that chooses a single index to perform queries on based on a few heuristics
+without comparing alternate paths:
```

Comment:
Why not just SQL queries directly? (Assuming we can ensure timeouts on request, and that its a read-only query within the APIs)

Reply:
That could be a possible frontend API or part of it but we'd still need to write a query planner to target the ORM data directly

Reply:
Hrmm, maybe I'm misunderstanding what GORM's SQL Builder guarantees is available (https://gorm.io/docs/sql_builder.html) , I thought the scan lines looked like what would suffice.

I'll try to look through more of the GORM examples to better understand!

Reply:
This has nothing to do with GORM. It's about the SDK's ORM. Different thing